mat64/list_test.go

```diff
@@ -0,0 +1,706 @@
+		if upper {
+			for i := 0; i < m; i++ {
+				for j := i; j < n; j++ {
+					mat.SetTri(i, j, rand.Float64())
+				}
+			}
+		} else {
+			for i := 0; i < m; i++ {
+				for j := 0; j <= i; j++ {
+					mat.SetTri(i, j, rand.Float64())
+				}
+			}
+		}
+		return returnAs(mat, t)
+	}
+}
+
+// makeCopyOf makes a copy of the matrix.
+func makeCopyOf(a Matrix) Matrix {
+	switch t := a.(type) {
+	default:
+		panic("unknown type in makeCopyOf")
+	case *Dense, *basicMatrix, *basicVectorer:
+		var m Dense
+		m.Clone(a)
+		return returnAs(&m, t)
+	case *SymDense, *basicSymmetric:
+		n := t.(Symmetric).Symmetric()
+		m := NewSymDense(n, nil)
+		m.CopySym(t.(Symmetric))
+		return returnAs(m, t)
+	case *TriDense, *basicTriangular:
+		n, upper := t.(Triangular).Triangle()
+		m := NewTriDense(n, upper, nil)
+		if upper {
+			for i := 0; i < n; i++ {
+				for j := i; j < n; j++ {
+					m.SetTri(i, j, t.At(i, j))
+				}
+			}
+		} else {
+			for i := 0; i < n; i++ {
+				for j := 0; j <= i; j++ {
+					m.SetTri(i, j, t.At(i, j))
+				}
+			}
+		}
+		return returnAs(m, t)
+	case *Vector:
+		m := &Vector{
+			mat: blas64.Vector{
+				Inc:  t.mat.Inc,
+				Data: make([]float64, len(t.mat.Data)),
+			},
+			n: t.n,
+		}
+		copy(m.mat.Data, t.mat.Data)
+		return m
+	case Transpose:
+		matrix := t.Matrix
+		mat := matrix.(*Vector) // Should only be here for the vector type
+		m := &Vector{
+			mat: blas64.Vector{
+				Inc:  mat.mat.Inc,
+				Data: make([]float64, len(mat.mat.Data)),
+			},
+			n: mat.n,
+		}
+		copy(m.mat.Data, mat.mat.Data)
+		return Transpose{m}
+	}
+}
+
+// makeTransposeCopyOf makes a transpose copy of the given matrix with the same
+// underlying type.
+func makeTransposeCopyOf(a Matrix) Matrix {
+	switch t := a.(type) {
+	default:
+		panic("unknown type in make transpose of")
+	case *Dense, *basicMatrix, *basicVectorer:
+		r, c := a.Dims()
+		m := NewDense(c, r, nil)
+		for i := 0; i < r; i++ {
+			for j := 0; j < c; j++ {
+				m.Set(j, i, a.At(i, j))
+			}
+		}
+		return returnAs(m, t)
+	case *SymDense, *basicSymmetric:
+		return makeCopyOf(a)
+	case *TriDense, *basicTriangular:
+		n, upper := t.(Triangular).Triangle()
+		mUpper := !upper
+		m := NewTriDense(n, mUpper, nil)
+		if mUpper {
+			for i := 0; i < n; i++ {
+				for j := i; j < n; j++ {
+					m.SetTri(i, j, t.At(j, i))
+				}
+			}
+		} else {
+			for i := 0; i < n; i++ {
+				for j := 0; j <= i; j++ {
+					m.SetTri(i, j, t.At(j, i))
+				}
+			}
+		}
+		return returnAs(m, t)
+	case *Vector:
+		m := makeCopyOf(a)
+		return Transpose{m}
+	case Transpose:
+		v := t.Matrix.(*Vector)
+		m := makeCopyOf(v)
+		return m
+	}
+}
+
+// sameType returns true if a and b have the same underlying type.
+func sameType(a, b Matrix) bool {
+	if reflect.ValueOf(a).Type() == reflect.ValueOf(b).Type() {
+		return true
+	}
+	return false
+}
+
+// canShadow returns true if the two matrices could be represented by the same
+// pointer.
+func canShadow(receiver, a Matrix, trans bool) bool {
```

Comment:
There have been some bugs discovered when performing operations where the receiver is the same as one of the input arguments, i.e. a.Mul(a,b). "canShadow" determines if it's possible with this combination of underlying Matrix types and sizes for that to be true. In other words, for multiply, if all the types are dense, and it's a 4x4 times a 4x4, it could be of the form
````
a.Mul(a,a)
a.Mul(a,b)
a.Mul(b,a)
c.Mul(a,b)
````
If instead b is a Symmetric, then only
````
a.Mul(a,b)
c.Mul(a,b)
````
are possible.